<a href="https://colab.research.google.com/github/ZackyRioOrlando/Praktikum-Pembelajaran-Mesin-/blob/main/JS03/JS03_TugasLab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

🧬
Tugas Lab

Wisconsin Breast Cancer (Again)

Deskripsi Tugas
Pada tugas pratikum ini Anda akan menggunakan data "Wisconsin Breast Cancer". Data tersebut terdiri dari 569 data yang digunakan untuk mendiagnonis jenis kanker Malignant (M) dan Benign (B). Tugas Anda adalah,

1. Pisahkan antara variabel yang dapat digunakan dan variabel yang tidak dapat digunakan.
2. Lakukan proses encoding pada kolom "diagnosis".
3. Lakukan proses standardisasi pada semua kolom yang memiliki nilai numerik.
4. Lakukan proses seleksi fitur. Anda dapat menggunakan SelectKBest.
5. Lakukan proses pengujian dengan model Logistic Regression seperti pada praktikum 1.
6. Anda dapat menggunakan model pipeline untuk mempermudah perkejaan Anda.
7. Berdasarkan hasil analisa Anda, berapa jumlah fitur terbaik yang dapat digunakan? Apa saja fitur tersebut?

Persiapan Data

In [18]:
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Memuat Data

In [20]:
from google.colab import files
uploaded = files.upload()

Saving wbc (1).csv to wbc (1) (1).csv


In [21]:
df = pd.read_csv('wbc (1).csv')
df.head()

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,...,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,NaN


Pengecekan Tipe Data

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

1. Pisahkan antara variabel yang dapat digunakan dan variabel yang tidak dapat digunakan.

In [22]:
# Menghapus kolom yang tidak digunakan
df = df.drop(columns=['id', 'Unnamed: 32'])

print("Jumlah variabel yang dapat digunakan:", df.shape[1])

print("\nVariabel yang dapat digunakan:")
print(df.columns.tolist())

Jumlah variabel yang dapat digunakan: 31

Variabel yang dapat digunakan:
['diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean', 'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean', 'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se', 'fractal_dimension_se', 'radius_worst', 'texture_worst', 'perimeter_worst', 'area_worst', 'smoothness_worst', 'compactness_worst', 'concavity_worst', 'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst']


2. Lakukan proses encoding pada kolom "diagnosis".

In [24]:
le = LabelEncoder() # membuat objek dari LabelEncoder
df['diagnosis'] = le.fit_transform(df['diagnosis']) # proses encoding
print(df['diagnosis'].unique())

[1 0]


In [8]:
df.head()

,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,...,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst
0,1,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,...,25.38,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890
1,1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,...,24.99,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902
2,1,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,...,23.57,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758
3,1,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,...,14.91,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300
4,1,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,...,22.54,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678


3. Lakukan proses standardisasi pada semua kolom yang memiliki nilai numerik.

In [9]:
# Mengambil semua kolom numerik
kolom_numerik = df.select_dtypes(include=['int64', 'float64']).columns

# Membuat objek StandardScaler
scaler = StandardScaler()

# Melakukan standardisasi
df[kolom_numerik] = scaler.fit_transform(df[kolom_numerik])

# Menampilkan hasil
print("Kolom yang distandardisasi:")
print(kolom_numerik.tolist())

print("\nHasil standardisasi:")
print(df.head())

Kolom yang distandardisasi:
['diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean', 'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean', 'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se', 'fractal_dimension_se', 'radius_worst', 'texture_worst', 'perimeter_worst', 'area_worst', 'smoothness_worst', 'compactness_worst', 'concavity_worst', 'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst']

Hasil standardisasi:
   diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0   1.297676     1.097064     -2.073335        1.269934   0.984375   
1   1.297676     1.829821     -0.353632        1.685955   1.908708   
2   1.297676     1.579888      0.456187        1.566503   1.558884   
3   1.297676    -0.768909      0.253732       -0.592687  -0.764464   
4   1.297676     1.750297     -1.151816  

4. Lakukan proses seleksi fitur. Anda dapat menggunakan SelectKBest.

In [26]:
# Seleksi fitur dengan SelectKBest
selector_filter = SelectKBest(score_func=f_classif, k=5)

# Membuat pipeline seleksi fitur dan Logistic Regression
pipe_filter = Pipeline([
    ("prep", preprocess),
    ("sel", selector_filter),
    ("clf", LogisticRegression(max_iter=1000))
])

In [28]:
from sklearn.pipeline import Pipeline
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

# Memisahkan fitur dan target
X = df.drop(columns=['diagnosis'])
y = df['diagnosis']

# Membuat pipeline preprocessing
preprocess = Pipeline([
    ("scaler", StandardScaler())
])

# Feature selection
selector_filter = SelectKBest(score_func=f_classif, k=5)

# Pipeline
pipe_filter = Pipeline([
    ("prep", preprocess),
    ("sel", selector_filter),
    ("clf", LogisticRegression(max_iter=1000))
])

5. Lakukan proses pengujian dengan model Logistic Regression seperti pada praktikum 1.

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    stratify=y,
    test_size=0.2,
    random_state=42
)

pipe_filter.fit(X_train, y_train)

pred = pipe_filter.predict(X_test)

print("=== Filter (ANOVA) + LR ===")
print("Accuracy:", accuracy_score(y_test, pred))
print(classification_report(y_test, pred))

=== Filter (ANOVA) + LR ===
Accuracy: 0.9649122807017544
              precision    recall  f1-score   support

           0       0.96      0.99      0.97        72
           1       0.97      0.93      0.95        42

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114



In [30]:
	# 1) Nama fitur setelah preprocess
	feat_names = pipe_filter.named_steps["prep"].get_feature_names_out()
	print("Nama fitur:", feat_names)
	print("\n")

	# 2) Mask & skor fitur terpilih (SelectKBest)
	sel = pipe_filter.named_steps["sel"]
	mask = sel.get_support()
	selected_names = feat_names[mask]
	selected_scores = sel.scores_[mask]
	top = sorted(zip(selected_names, selected_scores), key=lambda t: t[1], reverse=True)[:10]
	print("Top fitur:", top)

Nama fitur: ['radius_mean' 'texture_mean' 'perimeter_mean' 'area_mean'
 'smoothness_mean' 'compactness_mean' 'concavity_mean'
 'concave points_mean' 'symmetry_mean' 'fractal_dimension_mean'
 'radius_se' 'texture_se' 'perimeter_se' 'area_se' 'smoothness_se'
 'compactness_se' 'concavity_se' 'concave points_se' 'symmetry_se'
 'fractal_dimension_se' 'radius_worst' 'texture_worst' 'perimeter_worst'
 'area_worst' 'smoothness_worst' 'compactness_worst' 'concavity_worst'
 'concave points_worst' 'symmetry_worst' 'fractal_dimension_worst']


Top fitur: [('concave points_worst', np.float64(733.7249325739482)), ('perimeter_worst', np.float64(717.2464871041376)), ('radius_worst', np.float64(692.861395260564)), ('concave points_mean', np.float64(684.5268452011383)), ('perimeter_mean', np.float64(548.4132358502825))]


6. Anda dapat menggunakan model pipeline untuk mempermudah perkejaan Anda.

7. Berdasarkan hasil analisa Anda, berapa jumlah fitur terbaik yang dapat digunakan? Apa saja fitur tersebut?

Kita dapat mengetahui akurasi 96% didapatkan dari fitur
1. concave points_worst
2. perimeter_worst
3. radius_worst
4. concave points_mean
5. perimeter_mean

dan itulah 5 fitur terbaik yang dapat digunakan.